Fix the output of the boolean knob

Keraito opened the pull request:
Issue: While doing some work on fixing [an issue with the button knob discussed on Slack](https://storybooks.slack.com/archives/C55DBHZL0/p1526589520000555), I found out that the `boolean` knob's behaviour was also kind of broken: after the initial toggle the value of the knob would always return `undefined`. This was due to the incoming `change` object at https://github.com/storybooks/storybook/blob/master/addons/knobs/src/registerKnobs.js#L18 always being of the form: `{name: "name-of-knob", type: "boolean"}`, without a value. Easiest reproducible by adding a story with a boolean knob and just printing out its direct value.

## What I did
Change the behaviour when the incoming `change` object is of type `boolean`, to flip the current value.

Keraito committed:
Fix the output of the boolean knob

wuweiweiwu commented on addons/knobs/src/registerKnobs.js:
```diff
@@ -15,12 +15,16 @@ function setPaneKnobs(timestamp = +new Date()) {
 
-  knobOptions.value = value;
+  if (type === 'boolean') {
+    knobOptions.value = !knobOptions.value;
```
what if the default value is `false` and switched to `true` but since `value` is `undefined`, this will set `knobOptions.value` as `false` which is the wrong value. 

I would suggest a `undefined` check.

Keraito replied:
> initial toggled to `false`

What do you mean exactly? That the default value is true and the user toggles it to false? That the default is true and the user toggled it to false, and that toggle persisted after refresh? I'm kinda confused so could you elaborate more?

Keraito replied:
@wuweiweiwu in my change, I'm using the `knopOptions.value` stored in the current knob to calculate the new `knobOptions.value`. So i'm not using the `value` destructured from the incoming `change` object. That value is always `undefined`, so it's impossible to defer the correct boolean from that value. The knob itself initializes its `knobOptions.value` with the setted default value, so my change is basically proposing to flip that value every time, instead of relying on the incoming event value. Does this solve your issue? Sorry for the confusion.

Hypnosphi replied:
> That value is always `undefined`

I would prefer to fix this part

Keraito replied:
I will see what I can do

wuweiweiwu replied:
@Keraito ah my bad. I thought you were doing `!value` instead.

Keraito committed:
Revert "Fix the output of the boolean knob"

This reverts commit cdf9bb4ded31d30a86f131da9d3cb61d3bca416a.

Keraito committed:
Use the event instead of the ref to get the checkbox value + pure func components

Keraito committed:
Remove random brackets

Keraito commented:
> I would prefer to fix this part

The previous implementation used a ref in the input element and tried to access the value that way. That didn't work, so I removed the ref and used the `event.target.checked` value instead. Reverted the previous changes and seemed to work. 
Linter warned that the components should be written as a pure function, so I assumed this was the way. If not, please let me know.

Hypnosphi commented on addons/knobs/src/components/PropField.js:
```diff
@@ -23,23 +23,14 @@ const Label = styled('label')({
 
-export default class PropField extends React.Component {
-  onChange = e => {
-    this.props.onChange(e.target.value);
```
You probably need to change other knob types to pick `e.changes.value` now

Hypnosphi replied:
Nevermind, this one was actually unused

Hypnosphi commented:
@wuweiweiwu please review again

wuweiweiwu reviewed (approved):
LGTM 👍

Hypnosphi commented:
@Keraito boolean knob seems to work OK on release branch, so I removed the `patch` label
https://release-3-4--storybooks-official.netlify.com/?knob-Dollars=12.5&knob-Name=Storyteller&knob-Years%20in%20NY=9&knob-background=%23ffff00&knob-Age=70&knob-Items%5B0%5D=Laptop&knob-Items%5B1%5D=Book&knob-Items%5B2%5D=Whiskey&knob-Other%20Fruit=lime&knob-Birthday=1484870400000&knob-Nice=true&knob-Styles=%7B%22border%22%3A%223px%20solid%20%23ff00ff%22%2C%22padding%22%3A%2210px%22%7D&knob-Fruit=apple&selectedKind=Addons%7CKnobs.withKnobs&selectedStory=tweaks%20static%20values&full=0&addons=1&stories=1&panelRight=0&addonPanel=storybooks%2Fstorybook-addon-knobs

Hypnosphi commented:
Released as [`4.0.0-alpha.8`](https://github.com/storybooks/storybook/releases/tag/v4.0.0-alpha.8)
